### PR TITLE
ENH: temperature ffos

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

Apply Maggie's pmps database temperatures to her temperature ffos in the context of the common components.

Related links:
- https://github.com/pcdshub/lcls-twincat-pmps/blob/master/lcls-twincat-pmps/PMPS/AncillaryComponents/FB_TempSensor_FFO.TcPOU
- https://github.com/pcdshub/lcls-twincat-general/blob/master/LCLSGeneral/LCLSGeneral/POUs/Hardware/FB_TempSensor.TcPOU

More detailed info:

- Create a helper subclass of the temp sensor pmps block called `FB_CC_TempSensor`
- Replace all temperature handling with the new subclass
- Set fault thresholds to the current setpoint's temp setting when possible
- Veto it in the out state
- Include provisions for the SATT (2-4 blade stack style), which previously did not have pmps db states
- Include provisions for power slits, which previously did not have pmps db states
- For the slits, use the FBs from lcls-plc-kfe-motion, which are strictly more useful than the versions from lcls-plc-lfe-motion.
- Add tests for everything I added
- Add reset logic on veto + no autoreset

## Motivation and Context

We can prevent the beam from melting certain devices with temperature measurements.
The temperatures can be stored in the database.

Jira https://jira.slac.stanford.edu/browse/ECS-4472

## How Has This Been Tested?

- I wrote tests for every device and the base FB
- The tests are largely structured as:

  - Set the threshold via spoofing the PMPS DB in memory (start from https://github.com/pcdshub/lcls-twincat-motion/blob/master/lcls-twincat-motion/Library/Tests/Helpers/FB_PMPSJsonTestHelper.TcPOU and expand)
  - Set the IO readback temperature to way above or way below the threshold via pointer cheating
  - Confirm they react accordingly at various states while the states mover test moves things around

## Where Has This Been Documented?

This needs to be announced on release and on deploy. Deployment must be done carefully to avoid rolling over the FFO slots cap on each PLC.

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive comments
- [x] Test suite passes locally
- [x] Libraries are set to ``Always Newest`` version (``Library, *``)
- [x] Committed with ``pre-commit`` or ran ``pre-commit run --all-files``

### CODE REVIEW

File: lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_CC_TempSensor EXTENDS FB_TempSensor_FFO
+VAR_INPUT
+END_VAR
+VAR_OUTPUT
+END_VAR
+VAR
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[// Databse default read is zero if no entry
```
Comment (ZLLentz) [inline lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:13]: typo from past me

**Comment by ZLLentz** [comment]
This is done except for testing + slits
I intend to write unit tests if I can
I need to survey all the live slits deployments and see how different they are from the slits classes stored here
Also need to do a survey of other temperature sensor devices outside of this PR

**Comment by ZLLentz** [comment]
According to pcdsdevices, only PowerSlits and the ExitSlits have temperature sensors of the slits objects. I'll gloss over ExitSlits for now.
We have three power slits in happi:

1. sl1k0 (SL1K0:POWER) (lcls-plc-kfe-motion)
2. sl1l0 (SL1L0:POWER) (lcls-plc-lfe-motion)
3. sl2k0 (SL2K0:POWER) (lcls-plc-kfe-motion)

SL2L0 also exists even though it doesn't have a happi entry

In this repo we have:
FB_SLITS, which is blank (???)
FB_SLITS_POWER, which is fairly comprehensive.

SL1L0 and SL2L0 have essentially the same code, an FB_SLITS/FB_SLITS_POWER pair inside their local PLC code.
SL1K0 and SL2K0 have a similar deal going on

So, we nominally have 3 different version each of FB_SLITS and FB_SLITS_POWER that I need to untangle before I decide where to implement the temperature faults

**Comment by ZLLentz** [comment]
A diff of the kfe and lfe FB_SLITS reveals:

- kfe's has pmps inputs, an FF fb (not called here, delegated to FB_SLITS_POWER), and logging
- kfe looks newer overall

A diff of the kfe and lfe FB_SLITS_POWER reveals:
- kfe's has something called pmps mode
- kfe implements the fast fault on aperture size vs requested aperture size

A diff of the library and kfe FB_SLITS_POWER and kfe FB_SLITS reveals:

- there's basically no correlation between these
- the library power slits fb contains some combination of some parts of the code from the kfe library

so I think a proper update to this needs a lot of thought on the slits end

**Comment by ZLLentz** [comment]
This is remarkably close. On Monday I will test interactively to make sure I'm not missing something big, and after that I will request reviewers.

**Comment by ZLLentz** [comment]
This worked as I expected interactively
Requesting reviewers now

File: lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+<TcPlcObject Version="1.1.0.1">
+  <POU Name="FB_CC_TempSensor" Id="{62e52992-2558-4a9c-b1dd-b3b9616cbbcf}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_CC_TempSensor EXTENDS FB_TempSensor_FFO
+VAR
+    rtVetoReset: R_TRIG;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[// Database default read is zero if no entry
+bVeto S= fFaultThreshold = 0;
```
Comment (nrwslac) [inline lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:11]: What's the idea behind vetoing if this is 0?

Comment (ZLLentz) [reply lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:11]: 0 is the threshold we get if there is no entry in the database for reactive temperature.
Scientific consensus was that we should probably change the default number to something else but that's what it is for now
The idea was that we wouldn't force them to come up with temperatures for all the states, but maybe this is incorrect?

File: lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// Also reset the FFO for transitions to bVeto
+rtVetoReset(CLK:=bVeto);
+FFO.i_xReset := rtVetoReset.Q;
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:14]: Yes, this code resets the latching fault when we re-enter the veto (out) condition.

File: lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU (hunk 2 above)
Comment (ZLLentz) [reply lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:11]: Also, should I revise the comment maybe? I realize now my comment is unclear

Comment (nrwslac) [reply lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:11]: I was worried that there might be a use case where 0 is actually the threshold, but if its okay for now that works. The comment is okay.  I guess to me it's a bit counter-intuitive. If someone uses this FB they presumably want to fault given a threshold. If they aren't able to load one from the database shouldn't we fault?

Comment (ZLLentz) [reply lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:11]: It sounded like, in the future, 0 could potentially be a threshold for a system that needs to really really be cooled. At present, it's not a realistic fault temperature so I ignored this for now.

So, we could change the default to either something really really low (to fault if nothing was entered) or to something really really high (to ignore if nothing was entered).

I think there are good philosophical reasons for either which way.

**Review by nrwslac — DISMISSED** [review]
This looks good to me. I didn't see where the FB_CC_TempSensor does its database look up, but it must be in the fast fault implementation.

**Comment by ZLLentz** [comment]
I'm not going to merge this right away- the philosophy of "faulting if nothing is in the db" is worth pondering and getting other opinions on.

File: lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU (hunk 2 above)
Comment (ZLLentz) [reply lcls-twincat-common-components/Library/POUs/FB_CC_TempSensor.TcPOU:11]: After a few days of background-thread braining on this, I figure I should simplify the logic here to "trust what pmps gives me and treat it seriously". So, if PMPS says the temp threshold is 0, I treat it as zero.

Then, if I want different default behavior, I should solve this in lcls-twincat-pmps, where we can decide if "no temperature database entry" means "absolute zero" (always fault on any reading), or if "no entry" means "set the number really high" (so we never fault).

**Review by NSLentz — DISMISSED** [review]
I think it looks good nothing stands out to me besides the default behavior for a missing database threshold but I think we have already discussed this.

**Comment by ZLLentz** [comment]
- Removed the special behavior for temp 0
- Removed the unit test that tests for this special behavior

Now, the library here will simply trust the threshold given from pmps lib (which defaults to 0 right now).

**Pull request merged**